### DISCUSSION TITLE
(Cross-post) Looking for coding tool

### DISCUSSION THREAD

**Discussion by sgjohnson1981** [freeplane/freeplane#853] [General]
I posted this in r/learnprogramming. Thought I'd post here in case what I'd like to do is actually not difficult to implement in Freeplane already, or offer a feature request I guess.

Looking for recommendations for a coding tool. I've used Freeplane for some time and found you can set the view to "outline view." This transforms the view (without modifying the mindmap in any way) from a horizontal parent/child mindmap to a well...outline, with relationships still shown by connecting lines("edges" in Freeplane), but instead of child nodes connected to the right, they're underneath and indented to the right, just like lines of code. The nice thing is that importing and exporting code (just via copy/paste) from Freeplane maintains the relationships between nodes. In other words, when you copy and paste indented code straight out of an IDE or editor, Freeplane understands indented lines represent a child node, and copy and pasted code from Freeplane leaves the indentations in place. This also leaves in place the option of "collapsing" top level lines of code when working in Freeplane, just like you would be able to in any IDE. I also like using "connectors" (more Freeplane-speak for extraneous lines you can use to show relationships in addition to child/node) that can show code logic or function calls. The only drawback is the connector lines in outline view default to a particular style that don't lend themselves to following the logic, and you can't change the style or even move them. Sorry, I know I may be talking gibberish to anyone not familiar with Freeplane. You'd just have to fiddle with for a minute to see what I mean. The link below talks about something called "call graphs," which seems to be closest thing to what I'm envisioning, but I don't want to "output" anything. I'd just like to be able to mark up the code in-place in a graphical way that makes it easier to understand what's going on, then recall what's going on just at a glance.

https://github.com/koknat/callGraph

The other drawback to this is that you can't mix views. I can't arrange left to right multiple source code files as separate top to bottom outlines in the same mindmap, which would be intuitive for me since source code files are tabbed along the top of the editor in VS Code, which is the IDE I'm using. On a widescreen monitor it would be fantastic for drawing connections between lines of code in separate source code files, for essentially a "whole program" view.

Anyone figured out how to implement anything similar to what I'm looking for?

(Note: googling didn't find much. This link was the most relevant I could find.

https://betterprogramming.pub/a-technique-to-inspect-complex-code-with-mind-maps-2dc83bb4c3a9)

**Comment by euu2021** [comment]
Interesting! I only started to learn programming, but I noticed that pasting code into Freeplane makes it easier to understand it. In the IDE I feel lost when the code gets more complex 😅

> The only drawback is the connector lines in outline view default to a particular style that don't lend themselves to following the logic, and you can't change the style or even move them.

> The other drawback to this is that you can't mix views.

Now, you can use, in the normal view, a layout that is similar to the outline, so the connectors can be used normally. And, it's possible to mix layouts. 

See, for example, #641 and #819

**Reply by sgjohnson1981** [reply]
After I posted, I noticed the entry in the changelog about new layouts for 1.10.5. I'll have to see if the "first child with parent" child node alignment is similar enough to how code looks for my eyes. But the new fishbone in 1.11.1 is perfect for the second drawback I mentioned of not being able to see multiple source code files at one time. Thanks for the heads-up!!!

**Comment by EdoFro** [comment]
@sgjohnson1981 
I have some scripts to do something similar with VBA macros that I create for MS excel.

I export the .bas modules as files and copy the text as notes in the nodes.
Each module is copied in one node's note and has multiple functions and sub (methods).

1. a groovy scripts divides a module node in mutliple methods nodes (as its children)
2. other groovy script connects the different methods nodes showing how they call eachother
3. the rest of scripts are used to navigate between them (showing or hiding the ones related to the selected node)

here is a map with the scripts. It has no explanation, because it is for my own use, but it may be from some help to you.
[VBA example.zip](https://github.com/freeplane/freeplane/files/10123660/VBA.example.zip)


it transform a branch like this:
![image](https://user-images.githubusercontent.com/35700575/204809441-467a2972-e1f1-4434-b8cb-ed0e115af80b.png)

into this:
![image](https://user-images.githubusercontent.com/35700575/204809809-ad5089b7-496b-4624-b83d-765f2c9a9b8f.png)

best regards,
edo
